### ISSUE TITLE
vl_nnrelu much slower than vl_nnconv, strange behaviour

### ISSUE THREAD

**Issue by rosenfeldamir** [vlfeat/matconvnet#336]
Hi,
I've been profiling the run-time of training (specifically, imagenet vgg-f with batch normalization), after finding that the training can process "only" about 30 images / second. I expected it to be faster. I'm using a Tesla K80 GPU. 
According to the matlab profiler, most of the time is spent on `vl_nnrelu` : roughly 0.5s on a batch of size 256.  This is very different from what I'd expect, i.e, that `vl_nnconv` would take up a large portion of the time. The time spent on `vl_nnconv` is negligible in comparison : approx. 0.01s. 
The line taking up the run-time inside `vl_nnrelu` is of course the `y=max(x,single(0)) `
Specifically, here are the runtimes required by this operation along with the sizes of the inputs for the different relu's inside vgg-f:

54   54   64  256 : 0.076171 seconds
27   27  256  256 : 0.18188 seconds
13   13  256  256 : 0.092395 seconds
13   13  256  256 : 0.079397 seconds
13   13  256  256 : 0.079434 seconds
1     1  4096   256 : 0.015706 seconds
1     1  4096   256 : 0.006434 seconds 

Moreover, when I measure the runtime for a single gpuArray of size[27   27  256  256] , the results is much faster: 0.035 seconds. 
What is going on here? Why is the relu operation so "slow"? and does it make sense that the convolution is much faster?

**Comment by vedaldi** [comment]
Hi,

On slower hardware (Titan Black), we can train vgg-f  with about 280 images/s, so there is definitely a problem.

The profiler can be highly misleading as GPU operations in MATLAB are asynchronous. It may just happens that the GPU synchronises on the ReLU and so it *seems* that most of the time is spent there. You can try to debug that by running the network with the `sync` option on (this will be slower, but it should wait for the GPU after each command is scheduled).

I would also check how quickly you can read images from disk. For such network this is often a significant bottleneck. In particular, pre-processing the images to have a minimum side of 256 pixels and storing them on a fast disk (SSD or even a RAM disk if you have a lot of memory) might cure it.

Lastly, which platform are you using? We do all our large scale experiments in Linux.

**Comment by rosenfeldamir** [comment]
I'm using Linux : CentOS 6.6.
Indeed it seems that most of the time is spent on loading images using
vl_imreadjpeg, which is unreasonably slow despite the prefetching feature.
I've explored this in another issue which I think you will find
interesting,  #340

Thanks,
-Amir

**Comment by rosenfeldamir** [comment]
@vedaldi , when trying with the .sync option, things make more sense. However, the runtime is still considerably slower than reported for vgg-f. I am currently getting around 120Hz. This includes the time for loading the images.

**Comment by vedaldi** [comment]
Hi, ok, then the rest could be due to

* GPU model (our times are obtained with a Titan Black)
* Generally slow access to images

If you run your code with the profiler, can you see if the bottleneck is the vl_imreadjpeg / getBatch function or rather the CNN itself?

**Comment by rosenfeldamir** [comment]
Ok, I make a simple check, replacing the output of the get_batch function by an all-zeros array. The bottleneck is definitely due to loading the images from disk: Now I'm getting 200 images/sec , still slow compared to the Titan Black ( which is a bit surprising ).
In addition I've noticed that a considerable speedup can be obtained by switching from matlab's native  imresize to imResample from Piotr Dollar's toolbox:
http://vision.ucsd.edu/~pdollar/toolbox/doc/

**Comment by lenck** [comment]
Yeah, `imresize` is really slow (it does usually antialiasing as well...). That is why it is useful to pre-process the images with [this script](https://github.com/vlfeat/matconvnet/blob/master/utils/preprocess-imagenet.sh)... Plus, it is useful to have the pre-processed images on a ramdisk or at least SSD (preprocesed images from ILSVRC12 take approx. 56GB). If the images are of a correct size, the `imresize` is not being called at all.

**Comment by rosenfeldamir** [comment]
@lenck , the preprocessing surely speeds things up. However, It looks like the forward-backward pass is significantly slower than reported: for AlexNet, it requires roughly 1.8 seconds (using a simple tic/toc around the vl_simplenn call) for a batch of 256 images, i.e, about 140Hz. This half the speed I'd expect, given the info on the matconvnet website: if AlexNet does ~550Hz for forward passes, the forward-backward should be about 275Hz.  Do you know if anyone else has encountered this discrepancy?

**Comment by lenck** [comment]
In our case, we typically measure on Titan Black having all ILSVRC12 data preprocessed in a ramdisk. Usually, from my experience there is a big difference between gpu cards and also the speed of the storage matters a lot...
Also Matlab's profiler is really helpful to tackle the bottlenecks...

**Comment by rosenfeldamir** [comment]
Well, this happens on a Tesla K80. Vedaldi told me that the matlab profiler can be misleading when it comes to measuring gpu related performance. indeed this happened as it reported a long runtime on the RELU. So I resorted to a simple tic-toc method. I've also decoupled this from the speed of the storage.
